syntax = "proto2";

package raft;

message Entry {
  enum OperationType {
    PUT         = 1;
    GET         = 2;
    DIRECT_GET  = 3;
  }

  required int32 id           = 1;
  required OperationType type = 2;
  required string key         = 3;
  optional string value       = 4;
}

message AppendEntriesRequest {
  required uint64 term = 1;
  required uint64 leaderID = 2;
  required uint64 prevLogIndex = 3;
  required int64 prevLogTerm = 4;
  repeated Entry entries = 5;
  required uint64 leaderCommit = 6;
}

message AppendEntriesReply {
  required uint64 term = 1;
  required bool success = 2;
}

message RequestVoteRequest {
  required uint64 term = 1;
  required uint64 candidateID = 2;
  required uint64 lastLogIndex = 3;
  required uint64 lastLogTerm = 4;
}

message RequestVoteReply {
  required uint64 term = 1;
  required bool voteGranted = 2;
}
